geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/SSLDUnitTest.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class SSLDUnitTest {
+
+  @Rule
+  public RedisClusterStartupRule cluster = new RedisClusterStartupRule();
+
+  private static String commonPassword = "password";
+  private static SANCapturingHostnameVerifier hostnameVerifier = new SANCapturingHostnameVerifier();
+
+  private int redisPort;
+  private CertificateMaterial ca;
+  private String serverKeyStoreFilename;
+  private String serverTrustStoreFilename;
+
+  @Before
+  public void setup() throws Exception {
+    ca = new CertificateBuilder()
+        .commonName("Test CA")
+        .isCA()
+        .generate();
+
+    CertificateMaterial serverCertificate = new CertificateBuilder()
+        .commonName("server")
+        .issuedBy(ca)
+        .generate();
+
+    CertStores serverStore = CertStores.serverStore();
+    serverStore.withCertificate("server", serverCertificate);
+    serverStore.trust("ca", ca);
+    Properties serverProperties = serverStore.propertiesWith("all", true, false);
+
+    MemberVM server = cluster.startRedisVM(0, x -> x.withProperties(serverProperties));
+
+    redisPort = cluster.getRedisPort(server);
+    serverKeyStoreFilename = serverProperties.getProperty("ssl-keystore");
+    serverTrustStoreFilename = serverProperties.getProperty("ssl-truststore");
+  }
+
+  @Test
+  public void givenMutualAuthentication_clientCanConnect() throws Exception {
+    try (Jedis jedis = createClient(true, false)) {
+      assertThat(jedis.ping()).isEqualTo("PONG");
+    }
+  }
+
+  @Test
+  public void givenMutualAuthentication_clientErrorsWithoutKeystore() throws Exception {
+    IgnoredException.addIgnoredException(SSLHandshakeException.class);
+    IgnoredException.addIgnoredException("SunCertPathBuilderException");
+
+    Jedis jedis;
+    try {
+      // Create the client without a keystore
+      jedis = createClient(false, false);
+    } catch (JedisConnectionException ignored) {
+      return;
+    }
+
+    // Sometimes the client is created successfully - perhaps this is platform/JDK specific
+    assertThatThrownBy(jedis::ping).satisfiesAnyOf(
+        e -> assertThat(e.getMessage()).contains("SocketException"),
+        e -> assertThat(e.getMessage()).contains("SSLException"),
+        e -> assertThat(e.getMessage()).contains("SSLHandshakeException"));
+
+    IgnoredException.removeAllExpectedExceptions();
+  }
+
+  @Test
+  public void givenMutualAuthentication_clientErrorsWithSelfSignedCert() throws Exception {
+    IgnoredException.addIgnoredException(SSLHandshakeException.class);
+    IgnoredException.addIgnoredException("SunCertPathBuilderException");
+
+    Jedis jedis;
+    try {
+      // Create the client with a self-signed certificate
+      jedis = createClient(true, true);
+    } catch (JedisConnectionException ignored) {
+      return;
+    }
+
+    // Sometimes the client is created successfully - perhaps this is platform/JDK specific
```

Comment:
You might be able to write these tests a little more cleanly if you just wrap the both `createClient` and `ping` in one single assertThatThrownBy.

Reply:
Done.